### PR TITLE
feat(server): Implement GETEX command #385

### PR DESCRIPTION
**Pull request by matchyc**
Signed-off-by: matchyc <dawnlight.yc@protonmail.com>

issue: #385 
- Why we need this:
Add GETEX command
- What has been done:
Parsing args in `getex` command and using extended `OpGet` to finish it.

### CODE REVIEW

File: src/server/string_family.cc [hunk 1]
```diff
@@ -666,6 +666,91 @@ void StringFamily::GetSet(CmdArgList args, ConnectionContext* cntx) {
...
+    }
+    return;
+  } else {
+    cur_val = result.value();
+  }
+
+  SetCmd::SetParams sparams;
+  int64_t int_arg = 0;
+
+  for (size_t i = 2; i < args.size(); i++) {
```
Comment (romange) [inline src/server/string_family.cc:699]: unfortunately, this is not the right approach on how to implement commands in Dragonfly. 

This would break the command atomicity semantics because you have here 2 hops.
The right approach is to extend  `OpGet` functionality to support this use case and pass appropriate params there.

Please take a look at https://github.com/dragonflydb/dragonfly/pull/365 that had the same problem. Specifically:
1. Move argument parsing to the beginning of the function.
2. Introduce `ExpireParams` struct that consolidates all the expiry information. It should have a clear no-op/invalid state that means "no expiry info defined".
3. Extend OpGet to accept `ExpireParams` and optionally update expiry if it's defined. 

Some of the functionality above already exists in the codebase, so I took the liberty to introduce 
https://github.com/dragonflydb/dragonfly/pull/388 that should help you with implementing this PR. Please wait for it to be submitted and resync.

**Review by romange — CHANGES_REQUESTED** [review]

File: src/server/string_family.cc (hunk 1 above)
Comment (matchyc) [reply src/server/string_family.cc:699]: Thanks a lot! Waiting for pr and then trying to fix the inappropriate implementation.

Comment (romange) [reply src/server/string_family.cc:699]: The pr was merged, please rebase

File: src/server/string_family.cc [hunk 2]
```diff
@@ -666,6 +688,74 @@ void StringFamily::GetSet(CmdArgList args, ConnectionContext* cntx) {
...
+      return (*cntx)->SendError(kInvalidIntErr);
+    }
+
+    if (int_arg <= 0) {
+      return (*cntx)->SendError(InvalidExpireTime("getex"));
+    }
+
+    if (cur_arg == "EXAT" || cur_arg == "PXAT") {
+      exp_params.absolute = true;
+      int_arg = CalculateAbsTime(int_arg, is_ms);
```
Comment (romange) [inline src/server/string_family.cc:716]: `CalculateAbsTime` is an awful name because in fact, it calculates relative time. Long story short, you copied this parsing code from another place that does not use `ExpireParams` and that logic does not fit this specific struct.
you should pass `int_arg` as is without converting it here. 

Which brings me to another question - how the unit tests pass and whether you covered PXAT  well.

File: src/server/string_family.cc [hunk 3]
```diff
@@ -666,6 +688,74 @@ void StringFamily::GetSet(CmdArgList args, ConnectionContext* cntx) {
...
+      if (int_arg < 0) {
+        return (*cntx)->SendStored();
```
Comment (romange) [inline src/server/string_family.cc:718]: did you actually test this case with Redis? I will be very surprised that GETEX returns "OK" - it just does not make sense.

File: src/server/string_family.cc [hunk 4]
```diff
@@ -666,6 +688,74 @@ void StringFamily::GetSet(CmdArgList args, ConnectionContext* cntx) {
...
+      }
+    }
+
+    if (!is_ms && int_arg >= kMaxExpireDeadlineSec) {
```
Comment (romange) [inline src/server/string_family.cc:722]: please remove this check. it's wrong the way it lays here and `UpdateExpire` already checks the arguments for correctness.

File: src/server/generic_family.cc [hunk 5]
```diff
@@ -355,7 +355,6 @@ OpStatus OpExpire(const OpArgs& op_args, string_view key, const DbSlice::ExpireP
   auto [it, expire_it] = db_slice.FindExt(op_args.db_cntx, key);
   if (!IsValid(it))
     return OpStatus::KEY_NOTFOUND;
-
```
Comment (romange) [inline src/server/generic_family.cc:358]: please revert this change.

File: src/server/string_family_test.cc [hunk 6]
```diff
@@ -399,4 +399,58 @@ TEST_F(StringFamilyTest, GetDel) {
...
+TEST_F(StringFamilyTest, GetEx) {
+  using std::chrono::duration_cast;
+  using std::chrono::milliseconds;
+  using std::chrono::seconds;
+  using std::chrono::system_clock;
+
+  auto resp = Run({"set", "foo", "bar"});
+  ASSERT_THAT(resp, "OK");
+
+  // resp = Run({"getex", "foo", "EX", "0"});
```
Comment (romange) [inline src/server/string_family_test.cc:411]: why is it commented out?

File: src/server/string_family_test.cc [hunk 7]
```diff
@@ -399,4 +399,58 @@ TEST_F(StringFamilyTest, GetDel) {
...
+  // ASSERT_THAT(resp, ErrArg("invalid expire time"));
+
+  // resp = Run({"getex", "foo", "EX", "-1"});
+  // ASSERT_THAT(resp, ErrArg("invalid expire time"));
+
+  EXPECT_EQ(Run({"getex", "foo"}), "bar");
+
+  auto now = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
```
Comment (romange) [inline src/server/string_family_test.cc:419]: you should not use real clock. Instead use `TEST_current_time_ms` similarly to how it's done in 
`GenericFamilyTest.Expire`.

File: src/server/string_family_test.cc [hunk 8]
```diff
@@ -399,4 +399,58 @@ TEST_F(StringFamilyTest, GetDel) {
...
+  auto resp = Run({"set", "foo", "bar"});
+  ASSERT_THAT(resp, "OK");
+
+  // resp = Run({"getex", "foo", "EX", "0"});
+  // ASSERT_THAT(resp, ErrArg("invalid expire time"));
+
+  // resp = Run({"getex", "foo", "EX", "-1"});
+  // ASSERT_THAT(resp, ErrArg("invalid expire time"));
+
+  EXPECT_EQ(Run({"getex", "foo"}), "bar");
```
Comment (romange) [inline src/server/string_family_test.cc:417]: have you checked that the behavior modeled in the tests reflect the golden standard (aka Redis 6)?

did you actually tested these sequences with Redis 6?

**Review by romange — CHANGES_REQUESTED** [review]
Hi Meng,

Thanks for contributing to Dragonfly. I understand you are probably a junior engineer. I think this PR has room for improvement.

1. It seems you use copy-paste without actually checking whether the logic fits the use-case.
2. The tests should cover behavior according to how Redis responds, and it's not clear whether this was done. You should learn the feature with its corner cases by playing with Redis before you start coding it.

Tests are the most important part of a PR. Please note, I can not find bugs by just reviewing the code.

**Comment by matchyc** [comment]
> Hi Meng,
> 
> Thanks for contributing to Dragonfly. I understand you are probably a junior engineer. I think this PR has room for improvement.
> 
> 1. It seems you use copy-paste without actually checking whether the logic fits the use-case.
> 2. The tests should cover behavior according to how Redis responds, and it's not clear whether this was done. You should learn the feature with its corner cases by playing with Redis before you start coding it.
> 
> Tests are the most important part of a PR. Please note, I can not find bugs by just reviewing the code.

Thanks for the review! After improving this pr according to the suggestions, some obvious mistakes were fixed. 
One mess is the test case about EXAT, the timer for the "second level" expiration needs to be compatible with built-in `TEST_current_time_ms`, so I use several variables for the test case.

File: src/server/string_family_test.cc [hunk 9]
```diff
@@ -399,4 +399,80 @@ TEST_F(StringFamilyTest, GetDel) {
...
+  Run({"set", "foo", "bar"});
+
+  resp = Run({"getex", "foo", "EXAT", absl::StrCat(TEST_current_time_ms / 1000 - 1)});
+  EXPECT_EQ(resp, "bar");
+  EXPECT_THAT(Run({"getex", "foo"}), ArgType(RespExpr::NIL));
+
+  Run({"set", "foo", "bar"});
+
+  uint64_t next_two_seconds = TEST_current_time_ms + 2000;
+  uint64_t next_two_seconds_round_down = static_cast<uint64_t>(next_two_seconds / 1000.0);
```
Comment (romange) [inline src/server/string_family_test.cc:445]: nit: `next_two_seconds / 1000` will work as well.

File: src/server/string_family.cc [hunk 10]
```diff
@@ -666,6 +683,66 @@ void StringFamily::GetSet(CmdArgList args, ConnectionContext* cntx) {
...
+
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    return OpGet(t->GetOpArgs(shard), key, false, exp_params);
+  };
+
+  DVLOG(1) << "Before Get::ScheduleSingleHopT " << key;
+
+  OpResult<string> result = cntx->transaction->ScheduleSingleHopT(std::move(cb));
+
+  if (result.status() != OpStatus::OK) {
```
Comment (romange) [inline src/server/string_family.cc:731]: i would reverse the condition - just easier to read.
```cpp
if (result)
  return (*cntx)->SendBulkString(*result);

switch (result.status()) {
 ...
```

File: src/server/string_family.cc [hunk 11]
```diff
@@ -666,6 +683,66 @@ void StringFamily::GetSet(CmdArgList args, ConnectionContext* cntx) {
...
+    if (i >= args.size()) {
+      return (*cntx)->SendError(kSyntaxErr);
+    }
+
+    string_view ex = ArgS(args, i);
+    if (!absl::SimpleAtoi(ex, &int_arg)) {
+      return (*cntx)->SendError(kInvalidIntErr);
+    }
+
+    if (int_arg <= 0) {
```
Comment (romange) [inline src/server/string_family.cc:705]: looks much better!

question: why not implement the "persist" option? it's just a few more lines and you got the whole command covered.

**Review by romange — COMMENTED** [review]
Looks much, much better, thanks! And simpler, imho.

File: src/server/string_family.cc (hunk 11 above)
Comment (matchyc) [reply src/server/string_family.cc]: Now it may support `persist` option, checked by `TTL` command.

Comment (matchyc) [reply src/server/string_family.cc]: > looks much better!
> 
> question: why not implement the "persist" option? it's just a few more lines and you got the whole command covered.



> looks much better!
> 
> question: why not implement the "persist" option? it's just a few more lines and you got the whole command covered.

Now it may support persist option, checked by TTL command.

Comment (matchyc) [reply src/server/string_family.cc]: > looks much better!
> 
> question: why not implement the "persist" option? it's just a few more lines and you got the whole command covered.

Now it may support `persist` option, and the functionality was checked by `TTL` command.

**Review by romange — APPROVED** [review]
Looks great! Thanks again!
